### PR TITLE
Yield Farming Smart Contract

### PR DESCRIPTION
**Pull request by abancu**
In Progress
Added Interface for Stake

### CODE REVIEW

**Comment by rizedr** [comment]
[ch150]

File: test/YieldFarm.test.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        // await staking.manualEpochInit([usdc.address], 1)
+        const totalAmount = amount.mul(55).div(10)
+        expect(await yieldFarm.getPoolSize(1)).to.equal(totalAmount)
+        expect(await yieldFarm.getEpochStake(userAddr, 1)).to.equal(totalAmount)
+        expect(await barnBridge.balanceOf(yieldFarm.address)).to.equal(distributedAmount)
+        expect(await yieldFarm.getCurrentEpoch()).to.equal(3)
+        await yieldFarm.connect(user).harvest(1)
+        expect(await barnBridge.balanceOf(userAddr)).to.equal(distributedAmount.div(24))
+    })
+
```
Comment (kwix) [inline test/YieldFarm.test.js:87]: I would add a test with multiple users participating into the pools and also vary the pools.

File: contracts/YieldFarm.sol [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        mapping (address => bool) claimed;
+    }
+
+
+    // state variables
+    IERC20 private _barn;
+    address private _usdc;
+    address private _susd;
+    address private _dai;
+    address private _barnYCurve;
```
Comment (miladmostavi) [inline contracts/YieldFarm.sol:38]: Rename to _uni_lp please. We'll mostly use USDC_BOND_UNI_LP token.

File: contracts/YieldFarm.sol [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+
+    // modifiers
+
+
+
+
+    // construct
+
+    constructor(address barnBridgeAddress, address usdc, address susd, address dai, address barnYCurve, address stakeContract) public {
```
Comment (miladmostavi) [inline contracts/YieldFarm.sol:53]: barnBridgeAddress -> bondTokenAddress

Comment (miladmostavi) [reply contracts/YieldFarm.sol:53]: barnYCurve -> should not be part of the constructor -> should be calculated using the deterministic function most probably between _bond and _usdc UNI LP.

```/// https://uniswap.org/docs/v2/smart-contract-integration/getting-pair-addresses/
function genUniAddr(address left, address right) internal pure returns (address) {
    address first = left < right ? left : right;
    address second = left < right ? right : left;
    address pair = address(uint(keccak256(abi.encodePacked(
      hex'ff',
      uniFactory,
      keccak256(abi.encodePacked(first, second)),
      hex'96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f'
    ))));
    return pair;
}

File: contracts/YieldFarm.sol [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        require (_getEpochId() > epochId, "This epoch is in the future");
+
+        Epoch storage epoch = epochs[epochId];
+        require (epoch.claimed[msg.sender] != true, "User already claimed");
+
+        if (!epoch.init) { // first to harvest
+            epoch.poolSize = _getPoolSize(epochId); // Staking will throw if it is not initialized and cannot initialize
+            epoch.init = true;
+        }
+        // Give user interest
```
Comment (miladmostavi) [inline contracts/YieldFarm.sol:103]: Even in the comments: interest -> reward

File: contracts/YieldFarm.sol [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        uint userInterest = TOTAL_DISTRIBUTED_AMOUNT.mul(10**18).div(NR_OF_EPOCHS)
+                .mul(_getUserBalancePerEpoch(msg.sender, epochId))
+                .div(epoch.poolSize);
+        epoch.claimed[msg.sender] = true;
+        _barn.transfer(msg.sender, userInterest);
```
Comment (miladmostavi) [inline contracts/YieldFarm.sol:108]: We should `transferFrom(communityVault`

File: contracts/YieldFarm.sol [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        uint valueBarnYCurve = _staking.getEpochPoolSize(_barnYCurve, epochId);
+        valueBarnYCurve = computeBonus(valueBarnYCurve);
+        return valueUsdc.add(valueSusd).add(valueDai).add(valueBarnYCurve);
+    }
+
+    function _getUserBalancePerEpoch (address userAddress, uint epochId) internal view returns (uint){
+        uint valueUsdc = _staking.getEpochUserBalance(userAddress, _usdc, epochId);
+        uint valueSusd = _staking.getEpochUserBalance(userAddress, _susd, epochId);
+        uint valueDai = _staking.getEpochUserBalance(userAddress, _dai, epochId);
+        uint valueBarnYCurve = _staking.getEpochUserBalance(userAddress, _barnYCurve, epochId);
```
Comment (miladmostavi) [inline contracts/YieldFarm.sol:125]: `valueUniLP`

**Review by miladmostavi — CHANGES_REQUESTED** [review]

**Review by miladmostavi — APPROVED** [review]
👍

**Pull request merged**